android-json-form-wizard/src/main/java/com/vijay/jsonwizard/widgets/CountDownTimerFactory.java

```diff
@@ -0,0 +1,159 @@
+            millis = time * 60 * 1000;
+        }
+        interval = Long.parseLong(countdownInterval);
+    }
+
+    private void formatWidget(JSONObject jsonObject, Context context) {
+
+        String labelTextSize = jsonObject.optString(JsonFormConstants.LABEL_TEXT_SIZE, "8sp");
+        String labelTextColor = jsonObject.optString(JsonFormConstants.LABEL_TEXT_COLOR, "#535F67");
+        String progressBarBackgroundColor = jsonObject.optString(JsonFormConstants.PROGRESSBAR_BACKGROUND_COLOR, "#B6BBBE");
+        String progressBarColor = jsonObject.optString(JsonFormConstants.PROGRESSBAR_COLOR, "#535F67");
+        String progressBarTextColor = jsonObject.optString(JsonFormConstants.PROGRESSBAR_TEXT_COLOR, "#535F67");
+
+        labelView.setTextSize(FormUtils.getValueFromSpOrDpOrPx(labelTextSize, context));
+        labelView.setTextColor(Color.parseColor(labelTextColor));
+
+        progressBar.setBackgroundColor(progressBarBackgroundColor);
+        progressBar.setProgressColor(progressBarColor);
+        progressBar.setTextColor(progressBarTextColor);
+    }
+
+    private int getLayout() {
+        return R.layout.native_form_countdown_timer;
+    }
+
+    /**
+     * Count down to the end of the time specified
+     *
+     * @param millis            The time in milliseconds to count down to
+     * @param countdownInterval The intervals for running the countdown
+     */
+    private void startCountDown(final long millis, long countdownInterval, final Context context) {
+        new CountDownTimer(millis, countdownInterval) {
+            @Override
+            public void onTick(long millisUntilFinished) {
+                elapsedCount++;
+                progressBar.setText(getFormattedTimeText(millisUntilFinished));
+                int progress = (int) (elapsedCount * 100 / (millis / 1000));
+                progressBar.setProgress(progress);
+            }
+
+            @Override
+            public void onFinish() {
+                elapsedCount = 0;
+                progressBar.setText(getFormattedTimeText(0));
+                progressBar.setProgress(progressBarMaxValue);
+                ringAlarm(context);
+            }
+        }.start();
+    }
+
+    private void ringAlarm(Context context) {
+        Uri alert = RingtoneManager.getDefaultUri(RingtoneManager.TYPE_ALARM);
+        alarmTone = (alert == null) ? RingtoneManager.getRingtone(context, Settings.System.DEFAULT_RINGTONE_URI) : RingtoneManager.getRingtone(context, alert);
+        if (!alarmTone.isPlaying()) {
+            alarmTone.play();
+        }
+    }
+
+
+    private String getFormattedTimeText(long timeValue) {
+        return String.format(Locale.getDefault(), "%02d:%02d",
+                TimeUnit.MILLISECONDS.toMinutes(timeValue), TimeUnit.MILLISECONDS.toSeconds(timeValue));
+    }
+
+    /**
+     * Override this to provide more on countdown complete post processing
+     */
+    protected void onCountdownFinish() {
```

Comment:
Shouldn't this method be called in `onFinish` method of the `CountDownTimer`?

Reply:
Also, I would argue that `ringAlarm(context);` should be called here as part of post-processing. 

It's a custom action that a user would want to have control over or skip altogether.

Reply:
Yeah, I see the benefit to having the `ringAlarm(ctx)` in the overrideable post-processing method so that it's called `onFinish()`
Will refactor